src/liballoc/rc.rs

```diff
@@ -573,10 +573,11 @@ impl<T: ?Sized> Rc<T> {
     /// Constructs an `Rc` from a raw pointer.
     ///
     /// The raw pointer must have been previously returned by a call to a
-    /// [`Rc::into_raw`][into_raw].
+    /// [`Rc::into_raw`][into_raw] using the same `T`.
```

Comment:
I think this is not *quite* true, or at least is likely over-constraining. For example, I believe it is fine to go between the unsized representation and the sized representation via pointer casts (e.g,. `Rc<T>` unsizes to `Rc<dyn Debug>` and then you into_raw, cast to T, and from_raw).

I'm not sure what exact wording we want to guarantee here, though, maybe just same T is in practice both sufficient and good enough for most use cases. Maybe something like "layout and alignment equivalent to T; note that this is essentially a reference to reference transmute if the T differs from the one used by into_raw, and so the same conditions apply"?

Reply:
Yes, I thought about that as well, it could be done. My thinking, however, is that it's too flaky to try to document under which conditions this is safe (and therefore binds all future changes to `Rc`/`Arc`).

Reply:
How about this? (basically the same what @Mark-Simulacrum suggested)

> The raw pointer must have been previously returned by a call to a `Rc<U>::into_raw`, where `U` must have the same alignment and memory layout as `T` (which is trivially satisfied if `T == U`). Note that if `U` is not `T`, this is basically like transmuting references of different types. See `mem::transmute` for more information on what restrictions apply in that case.

I usually wouldn't mind over-constraining, but if we document this exactly, this PR is just a doc bug fix.